src/main/java/org/opensearch/ad/rest/handler/ModelValidationActionHandler.java

```diff
@@ -0,0 +1,718 @@
+                        DetectorValidationIssueType.CATEGORY,
+                        ValidationAspect.MODEL
+                    )
+                );
+            return;
+        }
+        BoolQueryBuilder query = QueryBuilders.boolQuery().filter(anomalyDetector.getFilterQuery());
+        for (Map.Entry<String, Object> entry : topEntity.entrySet()) {
+            query.filter(QueryBuilders.termQuery(entry.getKey(), entry.getValue()));
+        }
+        AggregationBuilder aggregation = getBucketAggregation(latestTime);
+        SearchSourceBuilder searchSourceBuilder = getSearchSourceBuilder(query, aggregation);
+        SearchRequest searchRequest = new SearchRequest(anomalyDetector.getIndices().toArray(new String[0])).source(searchSourceBuilder);
+        client.search(searchRequest, ActionListener.wrap(response -> processTopEntityResults(response, latestTime), listener::onFailure));
+    }
+
+    private void processTopEntityResults(SearchResponse response, long latestTime) {
+        Histogram aggregate = checkBucketResultErrors(response);
+        if (aggregate == null) {
+            return;
+        }
+        double fullBucketRate = processBucketAggregationResults(aggregate);
+        if (fullBucketRate < CONFIG_BUCKET_MINIMUM_SUCCESS_RATE) {
+            listener
+                .onFailure(
+                    new ADValidationException(
+                        CommonErrorMessages.CATEGORY_FIELD_TOO_SPARSE,
+                        DetectorValidationIssueType.CATEGORY,
+                        ValidationAspect.MODEL
+                    )
+                );
+        } else {
+            try {
+                checkFeatureQuery(latestTime);
+            } catch (Exception ex) {
+                logger.error(ex);
+                listener.onFailure(ex);
+            }
+        }
+    }
+
+    private void checkFeatureQuery(long latestTime) throws IOException {
+        ActionListener<MergeableList<double[]>> validateFeatureQueriesListener = ActionListener
+            .wrap(response -> { windowDelayRecommendation(latestTime); }, exception -> {
+                listener
+                    .onFailure(
+                        new ADValidationException(
+                            exception.getMessage(),
+                            DetectorValidationIssueType.FEATURE_ATTRIBUTES,
+                            ValidationAspect.DETECTOR
+                        )
+                    );
+            });
+        MultiResponsesDelegateActionListener<MergeableList<double[]>> multiFeatureQueriesResponseListener =
+            new MultiResponsesDelegateActionListener<>(
+                validateFeatureQueriesListener,
+                anomalyDetector.getFeatureAttributes().size(),
+                String.format(Locale.ROOT, CommonErrorMessages.VALIDATION_FEATURE_FAILURE, anomalyDetector.getName()),
+                false
+            );
+
+        for (Feature feature : anomalyDetector.getFeatureAttributes()) {
```

Comment:
This only checks each individual feature for causing the sparsity. What about the case of some combination of features is what causes the sparsity? For example, say feature `A` by itself passes the minimum success rate, and feature `B` by itself passes the minimum success rate, but when combining the queries of feature `A` and feature `B`, drops it below the minimum success rate. In that case, the API would return the features as being ok when it's actually the cause of data sparsity.

Suggest to combine all of these features into one query to holistically determine if the combination of features is the culprit of sparsity or not

Reply:
This is a fair point, original idea behind doing each feature at a time was to find if one of the feature is causing the issue. But as you say that might lead to feature query being skipped as root issue. I'll change to running them all at once with the only downside being that the user won't know which feature contributed more to the lack of sparsity unless I do further analysis which I wont be implementing at this point. Overall it does make sense to find the feature queries combined being the issue rather than saying there is no issue.

Reply:
Yeah, tradeoffs for either soln. Perfect soln would be trying to find a subset of features responsible for the sparsity, but even then, there could be multiple combinations, and could lead to a confusing response.

I don't have too strong opinion either way, but I slightly lean towards checking them altogether. That way it guarantees a valid response, even if it's more vague.